### ISSUE TITLE
Support dev-containers for better Codespaces experience

### ISSUE THREAD

**Issue by sobolevn** [HypothesisWorks/hypothesis#2735]
While working on #2729 I am trying out new GitHub's feature called "Codespaces": basically a vscode in your browser. https://github.com/features/codespaces

<img width="1280" alt="Снимок экрана 2021-01-04 в 13 04 33" src="https://user-images.githubusercontent.com/4660275/103524444-ab3a9d80-4e8e-11eb-9d56-acacf7870191.png">

It works fine with `hypothesis`, but there are little tweaks required:
- `shellcheck` is not installed
- tooling is not installed and configured by default
- some other?

We can configure it with dev-containers.
Docs: https://docs.github.com/en/free-pro-team@latest/github/developing-online-with-codespaces/configuring-codespaces-for-your-project#using-a-pre-built-container-configuration

Why is this great? It promises to simplify how other people can start contributing to our project.
Just one mouse click and you have a working dev environment.

Btw, can other devs access my editor? https://sobolevn-hypothesisworks-hypothesis-pv4w.github.dev/

**Comment by sobolevn** [comment]
One more thing: `git add --all` adds a lot of junk inside a codespace.

```
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   .github/workflows/main.yml
        new file:   .venv/bin/Activate.ps1
        new file:   .venv/bin/activate
        new file:   .venv/bin/activate.csh
        new file:   .venv/bin/activate.fish
```

**Comment by Zac-HD** [comment]
Sounds cool!  I've joined the waitlist and will try it out when I can 😄 

- I can't access your editor, and it looks like that's the intended behaviour
- Let's add `.venv` to our gitignore - we already have `venv` in there 😅 

If you're up for adding the `.devcontainer` and a sentence or two to the contributing guide that would be great, and I'll merge after I get access to the beta and can try it out!

**Comment by sobolevn** [comment]
@Zac-HD I will try to speed up the waitlist process 😉

**Comment by sobolevn** [comment]
@Zac-HD it should be available now!

**Comment by Zac-HD** [comment]
I'm in the beta!  Still can't access your codespace, which is a sensible default auth setting.  I imagine they'll add VSCode-style live collaboration once codespaces are out of beta 😁 

So I think the checklist is something like 

- [ ] Add `.venv` to `.gitignore`
- [ ] Add `.devcontainer` for custom codespace setup, including `shellcheck` and a Python that supports `sqlite` (I think the default Python3 container will have this, but on my trial run generic Ubuntu 20.04 didn't)
- [ ] Add a brief line in the contributor guide pointing to codespaces as an option

**Comment by sobolevn** [comment]
Quick update: I have enabled my first dev-container for `dry-python/returns`, feel free to check it out. Link: https://github.com/dry-python/returns/tree/master/.devcontainer

What it does:
- Pulls correct docker image
- Configures vscode with proper settings and extensions
- Installs `poetry` and project deps
- Configures your editor to work with this new env

Some screenshots.
<img width="1280" alt="Снимок экрана 2021-01-10 в 18 24 09" src="https://user-images.githubusercontent.com/4660275/104128007-34157580-5376-11eb-9482-e88096467fef.png">
<img width="1280" alt="Снимок экрана 2021-01-10 в 18 50 09" src="https://user-images.githubusercontent.com/4660275/104128010-37a8fc80-5376-11eb-9446-4117f4173bc2.png">

**Comment by sobolevn** [comment]
PR is incoming! 

P.S. Offtop, but I have promised to @Zac-HD  an article about how `hypothesis` helps us with "monad laws as values" somewhere in the dicussions (I cannot find the exact place of course). So! Here it is! https://sobolevn.me/2021/02/make-tests-a-part-of-your-app I hope that you will like the proposed idea 👍

**Comment by sobolevn** [comment]
Disussion about testing Codespaces: https://github.com/github/feedback/discussions/3464

TLDR: we can build and test a custom `Docker` image that would be a base one for our Codespace.
